File: 프로그래머스/3/152995. 인사고과/인사고과.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct {
    int a;
    int b;
} Score;

// 정렬 기준: a 내림차순, b 오름차순
int compare(const void* x, const void* y) {
    Score* s1 = (Score*)x;
    Score* s2 = (Score*)y;

    if (s1->a != s2->a)
        return s2->a - s1->a;   // 근무태도 내림차순
    return s1->b - s2->b;       // 동료평가 오름차순
}

// scores_rows는 2차원 배열 scores의 행 길이
int solution(int** scores, size_t scores_rows, size_t scores_cols) {
    int wanhoA = scores[0][0];
    int wanhoB = scores[0][1];
    int wanhoSum = wanhoA + wanhoB;

    Score* arr = (Score*)malloc(sizeof(Score) * scores_rows);
    for (size_t i = 0; i < scores_rows; i++) {
        arr[i].a = scores[i][0];
        arr[i].b = scores[i][1];
    }

    qsort(arr, scores_rows, sizeof(Score), compare);

    int maxB = 0;
    int rank = 1;

    for (size_t i = 0; i < scores_rows; i++) {
        int a = arr[i].a;
        int b = arr[i].b;

        // 지배당한 경우
        if (b < maxB) {
            if (a == wanhoA && b == wanhoB) {
                free(arr);
                return -1;
            }
            continue;
        }

        maxB = b;

        if (a + b > wanhoSum)
            rank++;
    }

    free(arr);
    return rank;
}

File: 프로그래머스/3/152995. 인사고과/인사고과.swift
import Foundation

func solution(_ scores: [[Int]]) -> Int {
    let wanhoA = scores[0][0]
    let wanhoB = scores[0][1]
    let wanhoSum = wanhoA + wanhoB

    // (a 내림차순, b 오름차순)
    let sorted = scores.sorted { x, y in
        if x[0] != y[0] { return x[0] > y[0] }
        return x[1] < y[1]
    }

    var maxB = -1
    var rank = 1

    for s in sorted {
        let a = s[0], b = s[1]

        // 이전에 a가 더 큰(또는 같은) 사람들 중 b가 더 큰 사람이 존재 => 지배됨
        if b < maxB {
            if a == wanhoA && b == wanhoB {
                return -1
            }
            continue
        }

        // 인센티브 대상인 경우만 석차 계산에 반영
        if a + b > wanhoSum {
            rank += 1
        }

        if b > maxB { maxB = b }
    }

    return rank
}